File: backend/app/schemas/emergency.py
from uuid import UUID
from datetime import datetime

from pydantic import (
    BaseModel,
    Field
)

from app.core.enums import (
    SeverityEnum,
    EmergencyStatusEnum,
    TransferStatusEnum
)

from app.schemas.common import (
    BaseSchema
)


# =========================================================
# CREATE EMERGENCY CASE
# =========================================================

class EmergencyCreate(BaseModel):

    facility_id: UUID

    medicine_id: UUID

    shortage_score: float = Field(
        ge=0
    )

    severity: SeverityEnum

    emergency_radius_km: int = Field(
        ge=1,
        le=100
    )

    required_quantity: int = Field(
        gt=0
    )


# =========================================================
# EMERGENCY RESPONSE
# =========================================================

class EmergencyResponse(BaseSchema):

    emergency_case_id: UUID

    facility_id: UUID
    facility_name: str

    medicine_id: UUID
    medicine_name: str

    shortage_score: float

    severity: SeverityEnum

    emergency_radius_km: int

    required_quantity: int

    available_quantity: int

    emergency_status: EmergencyStatusEnum

    triggered_at: datetime

    resolved_at: datetime | None = None

    # ---------------------------
    # AI Recommendation
    # ---------------------------

    transfer_id: UUID | None = None

    donor_facility: str | None = None

    transfer_status: TransferStatusEnum | None = None

    approved_quantity: int | None = None

    match_score: float | None = None

    transfer_distance_km: float | None = None

    cascade_safe: bool | None = None

    estimated_eta_minutes: int | None = None

    ai_reason: str | None = None

    alternative_donors:int | None = None
# =========================================================
# EMERGENCY MATCH RESPONSE
# =========================================================

class EmergencyMatchResponse(BaseSchema):

    match_id: UUID

    emergency_case_id: UUID

    source_facility_id: UUID

    medicine_id: UUID

    available_quantity: int

    transferable_quantity: int

    distance_km: float

    ranking_score: float

    is_selected: bool

File: frontend/src/types/emergency.ts
export type EmergencySeverity =
  | "WARNING"
  | "CRITICAL"
  | "EMERGENCY"

export type EmergencyStatus =
  | "ACTIVE"
  | "MATCHING"
  | "RESOLVED"

export interface Emergency {

    emergency_case_id: string

    facility_id: string
    facility_name: string

    medicine_id: string
    medicine_name: string

    shortage_score: number

    severity: string

    emergency_radius_km: number

    required_quantity: number

    available_quantity: number

    emergency_status: string

    triggered_at: string

    resolved_at: string | null

    transfer_id: string | null

    donor_facility: string | null

    transfer_status: string | null

    approved_quantity: number | null

    match_score: number | null

    transfer_distance_km: number | null

    cascade_safe: boolean | null

    estimated_eta_minutes: number | null

    ai_reason: string | null

    alternative_donors: number | null

    inventory_score?: number | null
    distance_score?: number | null
    risk_score?: number | null
    hospital_rating?: number | null
    patients_at_risk?: number | null
    delivery_confidence?: number | null
    incident_owner?: string | null
    assigned_dispatcher?: string | null
    incident_age_minutes?: number | null
    sla_remaining_minutes?: number | null
}